stake-pool/program/src/state.rs

```diff
@@ -0,0 +1,102 @@
+//! State transition types
+
+use crate::error::Error;
+use crate::instruction::{unpack, Fee};
+use solana_sdk::{entrypoint::ProgramResult, program_error::ProgramError, pubkey::Pubkey};
+use std::mem::size_of;
+
+/// Initialized program details.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct StakePool {
+    /// Owner authority
+    /// allows for updating the staking authority
+    pub owner: Pubkey,
+    /// Deposit authority nonce
+    /// for `create_program_address(&[state::StakePool account, "deposit"])`
+    pub deposit_nonce: u8,
+    /// Withdrawal authority nonce
+    /// for `create_program_address(&[state::StakePool account, "withdrawal"])`
+    pub withdraw_nonce: u8,
+    /// Pool Mint
+    pub pool_mint: Pubkey,
+    /// Owner fee account
+    pub owner_fee_account: Pubkey,
+    /// Pool token program id
+    pub token_program_id: Pubkey,
+    /// total stake under management
+    pub stake_total: u64,
+    /// total pool
+    pub pool_total: u64,
+    /// Fee applied to deposits
+    pub fee: Fee,
+}
+impl StakePool {
+    /// calculate the pool tokens that should be minted
+    pub fn calc_pool_amount(&self, stake_lamports: u64) -> Option<u128> {
+        (stake_lamports as u128)
+            .checked_mul((self.pool_total as u128).checked_add(1)?)?
+            .checked_div((self.stake_total as u128).checked_add(1)?)
```

Comment:
Maybe the pool could be initialized with a stake, this way you wouldn't need to do all of these adds.

Reply:
There is no way to guarantee stake_total will never be 0, might be simpler to just error out here though.

Reply:
In that case, you may prefer special logic for an empty pool or stake, where you default to 1:1 if either pool_total or stake_total is 0.  Otherwise do this calculation without the addition